server/src/main/java/org/opensearch/search/query/QueryPhaseSearcherWrapper.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.query;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.search.Query;
+import org.opensearch.common.util.FeatureFlags;
+import org.opensearch.search.aggregations.AggregationProcessor;
+import org.opensearch.search.internal.ContextIndexSearcher;
+import org.opensearch.search.internal.SearchContext;
+import org.apache.lucene.search.CollectorManager;
+
+import java.io.IOException;
+import java.util.LinkedList;
+
+/**
+ * Wrapper class for QueryPhaseSearcher that handles path selection for concurrent vs
+ * non-concurrent search query phase and aggregation processor.
+ *
+ * @opensearch.internal
+ */
+public class QueryPhaseSearcherWrapper implements QueryPhaseSearcher {
+    private static final Logger LOGGER = LogManager.getLogger(QueryPhaseSearcherWrapper.class);
+    private final QueryPhaseSearcher defaultQueryPhaseSearcher;
+    private final QueryPhaseSearcher concurrentQueryPhaseSearcher;
+
+    public QueryPhaseSearcherWrapper() {
+        this.defaultQueryPhaseSearcher = new QueryPhase.DefaultQueryPhaseSearcher();
+        this.concurrentQueryPhaseSearcher = FeatureFlags.isEnabled(FeatureFlags.CONCURRENT_SEGMENT_SEARCH)
+            ? new ConcurrentQueryPhaseSearcher()
+            : null;
+    }
+
+    /**
+     * Perform search using {@link CollectorManager}
+     *
+     * @param searchContext      search context
+     * @param searcher           context index searcher
+     * @param query              query
+     * @param hasTimeout         "true" if timeout was set, "false" otherwise
+     * @return is rescoring required or not
+     * @throws java.io.IOException IOException
+     */
+    @Override
+    public boolean searchWith(
+        SearchContext searchContext,
+        ContextIndexSearcher searcher,
+        Query query,
+        LinkedList<QueryCollectorContext> collectors,
+        boolean hasFilterCollector,
+        boolean hasTimeout
+    ) throws IOException {
+        if (concurrentQueryPhaseSearcher != null && searchContext.concurrentSegmentSearchEnabled()) {
```

Comment:
I believe the `concurrentQueryPhaseSearcher != null` is redundant, the `searchContext.concurrentSegmentSearchEnabled()` should never return `true` if `FeatureFlags.CONCURRENT_SEGMENT_SEARCH` is not set

Reply:
I think you're right but since we are explicitly setting it to null on L36 I figured it would be good to include a check just in case. I think we can leave this in until we remove the feature flag in the future?

Reply:
Here is the thing (why I don't like this `null`): if `searchContext.concurrentSegmentSearchEnabled()` returns true, we should take the concurrent search path, if `concurrentQueryPhaseSearcher` is for some reasons `null` - we have bug there.